base/src/WhisperStreamTransform.cpp

```diff
@@ -0,0 +1,180 @@
+		return false;
+	}
+
+	framemetadata_sp metadata = getFirstOutputMetadata();
+	FrameMetadata::FrameType frameType = metadata->getFrameType();
+	if (frameType != FrameMetadata::TEXT)
+	{
+		LOG_ERROR << "<" << getId() << ">::validateOutputPins input frameType is expected to be TEXT. Actual<" << frameType << ">";
+		return false;
+	}
+
+	return true;
+}
+
+void WhisperStreamTransform::addInputPin(framemetadata_sp& metadata, string& pinId)
+{
+	Module::addInputPin(metadata, pinId);
+	mDetail->mOutputMetadata = framemetadata_sp(new FrameMetadata(FrameMetadata::FrameType::TEXT));
+	mDetail->mOutputMetadata->copyHint(*metadata.get());
+	mDetail->mOutputPinId = addOutputPin(mDetail->mOutputMetadata);
+}
+
+bool WhisperStreamTransform::init()
+{
+	//intialize model
+	auto samplingStrategy = whisper_sampling_strategy::WHISPER_SAMPLING_GREEDY;
+	switch (mDetail->mProps.samplingStrategy)
+	{
+		case WhisperStreamTransformProps::DecoderSamplingStrategy::GREEDY:
+			samplingStrategy = whisper_sampling_strategy::WHISPER_SAMPLING_GREEDY;
+			break;
+		case WhisperStreamTransformProps::DecoderSamplingStrategy::BEAM_SEARCH:
+			samplingStrategy = whisper_sampling_strategy::WHISPER_SAMPLING_BEAM_SEARCH;
+			break;
+		default:
+			samplingStrategy = whisper_sampling_strategy::WHISPER_SAMPLING_GREEDY;
+	}
+	mDetail->mWhisperFullParams = whisper_full_default_params(samplingStrategy);
+	mDetail->mWhisperContextParams = whisper_context_default_params();
+	mDetail->mWhisperContext = whisper_init_from_file_with_params(mDetail->mProps.modelPath.c_str(), mDetail->mWhisperContextParams);
+	return Module::init();
+}
+
+bool WhisperStreamTransform::term()
+{
+	whisper_free_context_params(&mDetail->mWhisperContextParams);
+	whisper_free_params(&mDetail->mWhisperFullParams);
+	whisper_free(mDetail->mWhisperContext);
+	return Module::term();
+}
+
+bool WhisperStreamTransform::process(frame_container& frames)
+{
+	auto frame = frames.begin()->second;
+	sf::Int16* constFloatPointer = static_cast<sf::Int16*>(frame->data());
+	int numberOfSamples = frame->size() / 2;
+	for (int index = 0; index < numberOfSamples; index++) {
+		mDetail->inputAudioBuffer.push_back((float)constFloatPointer[index]/ 32768.0f);
+	}
+	if (mDetail->inputAudioBuffer.size() < mDetail->mProps.bufferSize) {
+		return true;
+	}
+	whisper_full(
+		mDetail->mWhisperContext,
+		mDetail->mWhisperFullParams,
+		mDetail->inputAudioBuffer.data(),
+		mDetail->inputAudioBuffer.size()
+	);
+	std::string output = "";
+	const int n_segments = whisper_full_n_segments(mDetail->mWhisperContext);
+	for (int i = 0; i < n_segments; ++i) {
+		const char* text = whisper_full_get_segment_text(mDetail->mWhisperContext, i);
+		output += text;
+	}
+	mDetail->inputAudioBuffer.clear();
+	auto outFrame = makeFrame(output.length());
+	memcpy(outFrame->data(), output.c_str(), output.length());
+	frames.insert(make_pair(mDetail->mOutputPinId, outFrame));
```

Comment:
I thought that frame accumulation should be in this module itself since there maybe different processes that need different audio source chunk sizes. Eg: Denoiser, VAD, Speaker Diarization etc. It would be better if these two modules are not tightly coupled.
This will also help in sources like FileReader etc.

Reply:
> So the output text is being paired with a very small end segment of the audio. It will be nicer if the output container has the audio samples along with it's text equivalent. Some ideas:
> 
>     * implement the sample accumulation in the source, we can use the same samplesize in source and xform. it is easier in the source because we can keep extending the frame without any memcpy
I think sample accumulation should take place in the transform for audio transform modules. Different audio transforms may need different chunk sizes. Eg: Denoising, VAD, Speaker Diarization etc. 
> 
>     * We can implement a module for Convert_16s32f_Sfs which one shot converts Int16 to FP32
Sure makes sense. I can raise a follow up issue.
> 
>     * Now in the Audio2Text Module you can send out Audio_Int16, Audio_FP32 and it's Text output
Ok I can make that change.

Reply:
Raised issue [](https://github.com/Apra-Labs/ApraPipes/issues/326)